### PR TITLE
Menu doesnt stay open on click

### PR DESCRIPTION
**Pull request by AGS-**
Made select and open have more typical behavior.

### CODE REVIEW

**Comment by AGS-** [comment]
Fixing errors brought up by travis

**Comment by AGS-** [comment]
@wincent I was trying to pass in the DOM element by using ref however I couldn't figure out a good way to do it since I also need to pass in the event via `this.handleClick.bind(this)` to do the comparison. Any way I could pass both?

File: src/components/ToolbarMenu.js [hunk 1]
```diff
@@ -14,54 +14,57 @@ import React, { PropTypes } from 'react';
...
+      <a
+        className="toolbar-menu toolbar-button"
+        onClick={this.handleOpen.bind(this)}
+        onMouseDown={preventDefault}
+        ref={() => {
+          document.addEventListener('click',
+            this.handleClick.bind(this),
+            false
+          );
+        }}
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: This isn't exactly how `ref` props are supposed to be used. The pattern is:

```
// In `render()`:
ref={node => this._node = node}
```

And then elsewhere (like in `componentDidMount()` or an event handler), access `this._node` instead of using `ReactDOM.findDOMNode`.

So your `document.addEventListener` should go in `handleOpen`, and your `handleClick` should be able to do an `if (this._node !== e.target)` or similar.

File: src/components/ToolbarMenu.js [hunk 2]
```diff
@@ -14,54 +14,57 @@ import React, { PropTypes } from 'react';
...
+  }
 
-   handleClick(e) {
-     if (ReactDOM.findDOMNode(this) !== e.target) {
-       e.preventDefault();
-       this.setState({ visibility: 'hidden' });
-     }
-   }
+  handleClick(e) {
+    // eslint-disable-next-line no-undef
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: You won't need this suppression if you don't use `ReactDOM` (and if you did have to use it, I think the right way to deal with the lint would be to `import` it).

File: src/components/ToolbarSelect.js [hunk 3]
```diff
@@ -29,25 +30,31 @@ export class ToolbarSelect extends React.Component {
...
     return (
       <a
         className="toolbar-select toolbar-button"
         onClick={this.handleOpen.bind(this)}
         onMouseDown={preventDefault}
         ref={() => {
-          document.addEventListener('click', this.handleClick.bind(this), false);
+          document.addEventListener('click',
+          this.handleClick.bind(this),
+          false);
```
Comment (wincent) [inline src/components/ToolbarSelect.js]: Similar issue here; I don't think this is the pattern we want to follow.

File: src/components/ToolbarSelect.js [hunk 4]
```diff
@@ -63,6 +70,7 @@ export class ToolbarSelect extends React.Component {
   }
 
   handleClick(e) {
+    // eslint-disable-next-line no-undef
```
Comment (wincent) [inline src/components/ToolbarSelect.js]: As above.

**Comment by wincent** [comment]
Awesome. Thanks for fixing this. I left a comment or four inline but I think this is looking close. Thanks!

File: src/components/ToolbarMenu.js (hunk 1 above)
Comment (AGS-) [inline src/components/ToolbarMenu.js]: I agree, that's a better pattern.

File: src/components/ToolbarMenu.js (hunk 2 above)
Comment (AGS-) [inline src/components/ToolbarMenu.js]: Haha yeah I felt bad about throwing the suppression in there

**Comment by AGS-** [comment]
@wincent Fixed!

File: src/components/ToolbarMenu.js [hunk 5]
```diff
@@ -14,27 +14,52 @@ import React, { PropTypes } from 'react';
...
+
+  render() {
+    const ulStyle = {
+      visibility: this.state.visibility
+    };
+    return (
+      <a
+        className="toolbar-menu toolbar-button"
+        onClick={this.handleOpen.bind(this)}
+        onMouseDown={preventDefault}
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: (Non-blocking feedback.)

I think we should either just kill the `preventDefault` function, or use it everywhere. It's a bit of a smell to have two ways of preventing a default.

File: src/components/ToolbarMenu.js [hunk 6]
```diff
@@ -14,27 +14,52 @@ import React, { PropTypes } from 'react';
...
+    if (this._node !== e.target) {
+      e.preventDefault();
+      this.setState({ visibility: 'hidden' });
+    }
+  }
+
+  handleOpen = e => {
+    e.preventDefault();
+    this.setState({ visibility: 'visible' });
+    document.addEventListener('click', this.handleClick.bind(this));
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: For hygiene, we should make sure we remove this event listener on unmount or close. The currently implementation will re-bind repeatedly.

File: src/components/ToolbarSelect.js [hunk 7]
```diff
@@ -14,42 +14,71 @@ import React, { PropTypes } from 'react';
...
+    (child, i) => {
+      if (!selectedChild || child.props.selected) {
+        selectedChild = child;
+      }
+      const onChildSelect =
       child.props.onSelect ||
-      onSelect && onSelect.bind(null, child.props.value, i);
-    return <ToolbarSelectOption {...child.props} onSelect={onChildSelect} />;
-  });
+      this.props.onSelect && this.props.onSelect.bind(null,
```
Comment (wincent) [inline src/components/ToolbarSelect.js]: Indentation here looks funky; line continuations should be indented 2 spaces.

File: src/components/ToolbarSelect.js [hunk 8]
```diff
@@ -14,42 +14,71 @@ import React, { PropTypes } from 'react';
...
+  }
 
-ToolbarSelect.propTypes = {
-  onSelect: PropTypes.func,
-  title: PropTypes.string,
-};
+  handleOpen = e => {
+    e.preventDefault();
+    this.setState({ visibility: 'visible' });
+    document.addEventListener('click', this.handleClick.bind(this));
```
Comment (wincent) [inline src/components/ToolbarSelect.js]: Same issue with unmount here. Should always have a `removeEventListener` to balance every `addEventListener` call.

File: src/components/ToolbarMenu.js [hunk 9]
```diff
@@ -14,27 +14,52 @@ import React, { PropTypes } from 'react';
...
-  label: PropTypes.string,
-};
+  constructor(props) {
+    super(props);
+    this.state = { visibility: 'hidden' };
+  }
+
+  render() {
+    const ulStyle = {
+      visibility: this.state.visibility
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: (Non-blocking feedback.)

I probably would have gone with a class here. It would result in a smaller change to the CSS file, and would mean we'd be more consistently using one method (predominantly class selectors) of applying presentation throughout the app.

**Comment by wincent** [comment]
Left a few small comments but this is very close now. Would be good to fix those, then get this in; we can always iterate more on it later if we need.

File: src/components/ToolbarMenu.js (hunk 9 above)
Comment (AGS-) [inline src/components/ToolbarMenu.js]: Alright, yeah I'll change this to use a css class instead.

File: src/components/ToolbarSelect.js (hunk 8 above)
Comment (AGS-) [inline src/components/ToolbarSelect.js]: Whoops!

**Comment by AGS-** [comment]
Alright, I'll update this PR in a bit.

**Comment by AGS-** [comment]
Alright, I made it use classname instead of style like we do in ToolbarButton. I also have it remove eventListeners now.

**Comment by AGS-** [comment]
graphiqlselectmenu

File: src/components/ToolbarMenu.js (hunk 5 above)
Comment (AGS-) [inline src/components/ToolbarMenu.js]: Ah forgot to put this in, let me do another commit really quick

File: src/components/ToolbarMenu.js [hunk 10]
```diff
@@ -40,7 +38,7 @@ export class ToolbarMenu extends React.Component {
         <svg width="14" height="8">
           <path fill="#666" d="M 5 1.5 L 14 1.5 L 9.5 7 z" />
         </svg>
-        <ul className="toolbar-menu-items" style={ulStyle}>
+        <ul className={'toolbar-menu-items' + (visible ? ' open' : '')}>
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: Could use the `classnames` package for this, but I think it's good to keep the dependency footprint small, so this is good.

File: src/components/ToolbarMenu.js [hunk 11]
```diff
@@ -50,13 +48,14 @@ export class ToolbarMenu extends React.Component {
   handleClick(e) {
     if (this._node !== e.target) {
       e.preventDefault();
-      this.setState({ visibility: 'hidden' });
+      this.setState({ visible: false });
+      document.removeEventListener('click', this.handleClick.bind(this));
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: Don't forget to also remove it, if present, in `componentWillUnmount`.

**Comment by wincent** [comment]
I think we're good to land this after the `componentWillUnmount` change that I commented on inline.

File: src/components/ToolbarMenu.js [hunk 12]
```diff
@@ -25,6 +25,10 @@ export class ToolbarMenu extends React.Component {
     this.state = { visible: false };
   }
 
+  componentWillUnmount() {
+    document.removeEventListener('click', this.handleClick.bind(this));
```
Comment (wincent) [inline src/components/ToolbarMenu.js]: I am not sure that any of these `removeEventListener` calls will do what you want, because when you call `bind` you get a new function that is not `===`, which means that you won't remove anything here. (I could be wrong about that, but that's what I think will happen.)

I think the way to do this is as follows:

```
_subscribe() {
  if (!this._listener) {
    this._listener = this.handleClick.bind(this);
    document.addEventListener('click', this._listener);
  }
}

_release() {
  if (this._listener) {
    document.removeEventListener('click', this._listener);
    this._listener = null;
  }
}
```

And then call those methods.

Comment (AGS-) [inline src/components/ToolbarMenu.js]: Ok, makes sense. Yeah I was afraid of that but I wasn't really sure. Implemented _subscribe and _release and tested that everything works.

**Comment by wincent** [comment]
Sweet!
